src/test/java/com/google/devtools/build/lib/server/GrpcServerTest.java

```diff
@@ -460,6 +469,320 @@ public void onCompleted() {
         .isEqualTo(Code.INTERRUPTED);
   }
 
+  /**
+   * Ensure that if a command is marked as preemptible, running a second command
+   * interupts the first command
+   */
+  @Test
+  public void testPreeempt() throws Exception {
+    String firstCommandArg = "Foo";
+    String secondCommandArg = "Bar";
+
+    CommandDispatcher dispatcher =
+        new CommandDispatcher() {
+          @Override
+          public BlazeCommandResult exec(
+              InvocationPolicy invocationPolicy,
+              List<String> args,
+              OutErr outErr,
+              LockingMode lockingMode,
+              String clientDescription,
+              long firstContactTimeMillis,
+              Optional<List<Pair<String, String>>> startupOptionsTaggedWithBazelRc)
+              throws InterruptedException {
+            if (args.contains(firstCommandArg)) {
+              OutputStream out = outErr.getOutputStream();
+              try {
+                while (true) {
+                  if (Thread.interrupted()) {
+                    return BlazeCommandResult.failureDetail(
+                        FailureDetail.newBuilder()
+                            .setInterrupted(
+                                Interrupted.newBuilder().setCode(Code.INTERRUPTED))
+                            .build());
+                  }
+                  out.write(new byte[1024]);
+                }
+              } catch (IOException e) {
+                throw new IllegalStateException(e);
+              }
+            } else {
+              return BlazeCommandResult.success();
+            }
+          }
+        };
+    createServer(dispatcher);
+
+    CountDownLatch gotFoo = new CountDownLatch(1);
+    AtomicReference<RunResponse> lastFooResponse = new AtomicReference<>();
+    AtomicReference<RunResponse> lastBarResponse = new AtomicReference<>();
+
+    CommandServerStub stub = CommandServerGrpc.newStub(channel);
+    stub.run(
+        createPreemptibleRequest(firstCommandArg),
+        new StreamObserver<RunResponse>() {
+          @Override
+          public void onNext(RunResponse value) {
+            gotFoo.countDown();
+            lastFooResponse.set(value);
+          }
+
+          @Override
+          public void onError(Throwable t) {}
+
+          @Override
+          public void onCompleted() {}
+        });
+
+    // Wait for the first command to startup
+    gotFoo.await();
+
+    CountDownLatch gotBar = new CountDownLatch(1);
+    stub.run(
+        createRequest(secondCommandArg),
+        new StreamObserver<RunResponse>() {
+          @Override
+          public void onNext(RunResponse value) {
+            gotBar.countDown();
+            lastBarResponse.set(value);
+          }
+
+          @Override
+          public void onError(Throwable t) {}
+
+          @Override
+          public void onCompleted() {}
+        });
+
+    gotBar.await();
+    server.shutdown();
```

Comment:
instead of relying on the side-effects of shutdown would it be cleaner to wait on the commands finishing (via latches in onError/onCompleted). I *think* this will be easier to follow than relying on the side-effect of shutdown to be an orderly shutdown.

Reply:
The shutdown mechanics are used throughout the other tests. Usually I'd try to keep things consistent. What's your philosophy on cleaning up the other tests as part of this change? Happy to fix them as well.

Reply:
I think it's fine to leave it as it is for consistency.